src/main/software/amazon/event/ruler/Range.java

```diff
@@ -56,53 +54,67 @@ private Range(Range range) {
     }
+
     public static Range lessThanOrEqualTo(final double val) {
         return new Range(-Constants.FIVE_BILLION, false, val, false);
     }
+
     public static Range greaterThan(final double val) {
         return new Range(val, true, Constants.FIVE_BILLION, false);
     }
+
     public static Range greaterThanOrEqualTo(final double val) {
         return new Range(val, false, Constants.FIVE_BILLION, false);
     }
+
     public static Range between(final double bottom, final boolean openBottom, final double top, final boolean openTop) {
         return new Range(bottom, openBottom, top, openTop);
     }
 
-    private static Range deepCopy(final Range range) { return new Range(range); }
+    private static Range deepCopy(final Range range) {
+        return new Range(range);
+    }
 
     /**
      * This is necessitated by the fact that we do range comparisons of numbers, fixed-length strings of digits, and
      *  in the case where the numbers represent IP addresses, they are hex digits.  So we need to be able to say
      *  "for all digits between '3' and 'C'". This is for that.
      *
      * @param first Start one digit higher than this, for example '4'
-     * @param last Stop one digit lower than this 'B'
-     * @return The digit list, for example [ '4, '5', '6', '7', '8', '9', '9', 'A' ] (with 'B' for longDigitSequence)
+     * @param last Stop one digit lower than this, for example 'B'
+     * @return The digit list, for example [ '4', '5', '6', '7', '8', '9', '9', 'A' ] (with 'B' for longDigitSequence)
      */
-    static List<Byte> digitSequence(byte first, byte last, boolean includeFirst, boolean includeLast) {
-        assert first <= last && first <= 'F'&& first >= '0'&& last <= 'F' && last >= '0';
+    static byte[] digitSequence(byte first, byte last, boolean includeFirst, boolean includeLast) {
+        assert first <= last && first <= 'F' && first >= '0' && last <= 'F';
         assert !((first == last) && !includeFirst && !includeLast);
 
-        final List<Byte> bytes = new ArrayList<>();
-        int i = 0;
-        while (Constants.HEX_DIGITS[i] < first) {
-            i++;
-        }
+        int i = getHexByteIndex(first);
+        int j = getHexByteIndex(last);
 
         if ((!includeFirst) && (i < (Constants.HEX_DIGITS.length - 1))) {
             i++;
         }
-        while (Constants.HEX_DIGITS[i] < last) {
-            bytes.add(Constants.HEX_DIGITS[i++]);
-        }
+
         if (includeLast) {
-            bytes.add(Constants.HEX_DIGITS[i]);
+            j++;
         }
 
+        byte[] bytes = new byte[j - i];
+
+        System.arraycopy(Constants.HEX_DIGITS, i, bytes, 0, j - i);
+
         return bytes;
     }
 
+    private static int getHexByteIndex(byte value) {
+        // ['0'-'9'] maps to [0-9] indexes
+        if (value >= 48 && value <= 57) {
+            return value - 48;
+        }
+        // ['A'-'F'] maps to [10-15] indexes
+        return (value - 65) + 10;
```

Comment:
Any reason not to just go with -55 here? If you're trying to lay out both steps here for understandability/readability, perhaps it would be better to go further and turn these numbers into private static final class variables?

Reply:
mostly to try to keep it easy to understand, if we trust the compiler this is being optimized away. 

I think the variable may be the better option then